services/device-registry-mongodb/src/test/java/org/eclipse/hono/deviceregistry/mongodb/model/MongoDbBasedDeviceDtoTest.java

```diff
@@ -0,0 +1,97 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.deviceregistry.mongodb.model;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+
+import org.eclipse.hono.deviceregistry.mongodb.utils.MongoDbDeviceRegistryUtils;
+import org.eclipse.hono.service.management.BaseDto;
+import org.eclipse.hono.service.management.device.Device;
+import org.eclipse.hono.service.management.device.DeviceStatus;
+import org.eclipse.hono.util.RegistryManagementConstants;
+import org.junit.jupiter.api.Test;
+
+import io.vertx.core.json.JsonObject;
+
+
+class MongoDbBasedDeviceDtoTest {
+
+    /**
+     * Tests that a DTO created for inserting a new entry is serialized to JSON as expected in order to verify that
+     * the document to be persisted is written as expected.
+     */
+    @Test
+    public void testEncodeForCreate() {
+        final String tenantId = "barfoo";
+        final String deviceId = "foobar";
+        final DeviceStatus deviceStatus = new DeviceStatus()
+                // make sure that these values do not interfere with the corresponding properties of DeviceDto's parent classes
+                .setAutoProvisioned(true)
+                .setCreationTime(Instant.now().plusSeconds(3600));
+        final Device device = new Device();
+        device.setStatus(deviceStatus);
+        final String version = "spam";
+
+        final var deviceDto = MongoDbBasedDeviceDto.forCreation(MongoDbBasedDeviceDto::new, tenantId, deviceId, true, device, version);
+        final var json = JsonObject.mapFrom(deviceDto);
+
+        assertThat(json).isNotNull();
+        // size should match number of assertions below, to make sure no unexpected properties are serialized
+        assertThat(json.getMap()).hasSize(7);
+
+        assertThat(json.getString(RegistryManagementConstants.FIELD_PAYLOAD_TENANT_ID)).isEqualTo(tenantId);
+        assertThat(json.getString(RegistryManagementConstants.FIELD_PAYLOAD_DEVICE_ID)).isEqualTo(deviceId);
+        assertThat(json.getInstant(RegistryManagementConstants.FIELD_STATUS_CREATION_DATE)).isNotNull();
```

Comment:
should this be related to the instant set in the `DeviceStatus`?

Reply:
what about the *updated on* field? Shouldn't that contain the same instant as the *created on* field in this case?

Reply:
> should this be related to the instant set in the DeviceStatus?

No, actually the device status should be ignored.  

I added a DeviceStatus object to the test since the properties of DeviceStatus partially have the same name as the properties in the DTO subclasses. Before adding the _@JsonIgnore_ annotation to the _deviceStatus_ property in _DeviceDto_ they were accidentally serialised but the error was hard to spot since everything was still working but there was an additional, unwanted property being serialized. 

To make sure this doesn't accidentally happen in future (e.g. when someone changes the Jackson annotation again), I added that to the test case as well.

> what about the updated on field? Shouldn't that contain the same instant as the created on field in this case?

If the DTO was created by using forCreation() it can't have been updated yet which is why there is no update timestamp set. We could also set the same timestamp for created and updated but it is more cumbersome for someone viewing the entry to spot whether it has been updated at all yet (one needs to compare the created and updated timestamp for equality then).

Reply:
> No, actually the device status should be ignored.
>
> I added a DeviceStatus object to the test since the properties of DeviceStatus partially have the same name as the properties in the DTO subclasses. Before adding the @JsonIgnore annotation to the deviceStatus property in DeviceDto they were accidentally serialised but the error was hard to spot since everything was still working but there was an additional, unwanted property being serialized.

That's a good idea :+1:  IMHO we should make it explicit then and add corresponding assertions as well.

Reply:
It's already there (if I got you right):

`// size should match number of assertions below, to make sure no unexpected properties are serialized
  assertThat(json.getMap()).hasSize(7);`

By that it is made sure there is no other property serialized besides the one for which the test case contains assertions.

Reply:
I was thinking about
```
// make sure that the creation date set on the new device is not the one contained in the DeviceStatus
assertThat(json.getInstant(RegistryManagementConstants.FIELD_STATUS_CREATION_DATE)).isBefore(deviceStatus.getCreationTime());
```